framework/src/modules/chain/submodules/blocks/process.js

```diff
@@ -427,11 +435,242 @@ class Process {
+				return this._handleDoubleForgingTieBreak(block, lastBlock);
+			}
+		} else if (
+			lastBlock.heightPrevoted < block.heightPrevoted ||
+			(lastBlock.height < block.height &&
+				lastBlock.heightPrevoted === block.heightPrevoted)
+		) {
+			// Case 5: received block has priority. Move to a different chain.
+
+			return this._handleMovingToDifferentChain();
+			// TODO: Move to a different chain
+		}
+
+		// Discard newly received block
+		return library.logger.warn(
+			`Discarded block that does not match with current chain: ${
+				block.id
+			} height: ${block.height} round: ${slots.calcRound(
+				block.height
+			)} slot: ${slots.getSlotNumber(block.timestamp)} generator: ${
+				block.generatorPublicKey
+			}`
+		);
+	}
+
+	/**
+	 * Block IDs are the same ~ Blocks are equal
+	 * @param block
+	 * @returns {*}
+	 * @private
+	 */
+	// eslint-disable-next-line class-methods-use-this
+	_handleSameBlockReceived(block) {
+		library.logger.debug('Block already processed', block.id);
+	}
+
+	/**
+	 * Block received is correct
+	 * @param block
+	 * @returns {Promise}
+	 * @private
+	 */
+	_handleGoodBlock(block) {
+		return this._processBlock(block);
+	}
+
+	/**
+	 * Double forging. Last block stands
+	 * @param block
+	 * @returns {*}
+	 * @private
+	 */
+	// eslint-disable-next-line class-methods-use-this
+	_handleDoubleForging(block) {
+		library.logger.warn(
+			'Delegate forging on multiple nodes',
+			block.generatorPublicKey
+		);
+		library.logger.info('Last block stands');
+		// TODO: Implement Proof of Misbehavior
+	}
+
+	/**
+	 * Tie break: two competing blocks by different delegates at the same height.
+	 * @param lastBlock
+	 * @param newBlock
+	 * @returns {Promise}
+	 * @private
+	 */
+	async _handleDoubleForgingTieBreak(newBlock, lastBlock) {
+		try {
+			// Normalize and check schema
+			const normalizedNewBlock = library.logic.block.objectNormalize(
+				_.cloneDeep(newBlock)
+			);
+
+			// Validate block has been generated by correct delegate
+			await this._validateGenerator(normalizedNewBlock, lastBlock);
+
+			// Verify block data, signatures, etc
+			// TODO: Maybe `verifyReceipt` could throw errors instead of returning checks.
+			const check = modules.blocks.verify.verifyReceipt(normalizedNewBlock);
+
+			if (!check.verified) {
+				library.logger.error(
+					`Block ${normalizedNewBlock.id} verification failed`,
+					check.errors.join(', ')
+				);
+				// Return first error from checks
+				throw check.errors[0];
+			}
+		} catch (error) {
+			// Should there be any errors while validating and verifying,
+			// last block is kept as the tip of the chain
+			library.logger.error(
+				'Case 4 fork recovery failed',
+				convertErrorsToString(error)
+			);
+			throw error;
+		}
+
+		// If the new block is correctly validated and verified,
+		// last bloc kis deleted and new block is added to the tip chain
+		library.logger.info('Deleting last block due to case 4');
+		await deleteLastBlock();
```

Comment:
The LIP says that we keep reference of the old block, and in case we can't process the new block, we restore the old one.

Reply:
Yea, Jan suggested creating a temporary table to store it in case something goes wrong.

Reply:
Do we need to create a table for this? If process block fails it will return an exception that can be caught and last block can be restored back again.

Reply:
Yes I agree we don't need store those blocks to table, just keep in memory. Even if node crashes, it can sync to network. There is no fatal, non-recoverable lost.

Reply:
To clarify this, didn't we decide on storing those blocks in a temporary table as that's more convenient?

Reply:
I think in this particular case were we only keep the backup of the last block in memory, maybe we don't need to do that. If something fails in between and last block is lost, it will sync again